crates/wit-parser/src/resolve.rs

```diff
@@ -1329,11 +1340,23 @@ impl Resolve {
         }
     }
 
-    fn include_stability(&self, stability: &Stability) -> bool {
-        match stability {
-            Stability::Stable { .. } | Stability::Unknown => true,
+    fn include_stability(&self, stability: &Stability, pkg_id: &PackageId) -> Result<bool> {
+        Ok(match stability {
+            Stability::Unknown => true,
+            Stability::Stable { since, .. } => {
+                match self.packages.get(*pkg_id) {
+                    // If the `@since` version specified on this stability is ahead of the version on the
+                    // package that it belongs to, we throw an explicit error as that is not allowed
+                    Some(p) if p.name.version.as_ref().is_some_and(|v| since > v) => {
+                        bail!("referring to an unreleased (future) package version in a feature gate is not allowed");
```

Comment:
Could this error message include the two versions in question as well? (since there's no span information here so this isn't pointing to WIT files themselves).

Also, should this return an error if the package's version is `None` perhaps?

Reply:
Thanks for the quick review! So we can definitely include the two versions -- I've updated the message a bit.

> Also, should this return an error if the package's version is None perhaps?

And nope -- some packages that come in may not have versions specified at all (which we don't yet consider a proper bug). 

In particular the following tests have this use case:
- https://github.com/bytecodealliance/wasm-tools/blob/main/tests/cli/wit-stability-in-binary-format.wit
- https://github.com/bytecodealliance/wasm-tools/blob/main/tests/cli/wit-stability-inherited.wit

Do we want this to be an error? Effectively encoding that people *must* use top-level versioned-package WITs if they want to use `@since`? Obviously `@unstable` does not really require the same.

Reply:
My thinking is that if we're going to so far as to check version ranges that we should probably check that versions are present at all. I agree some tests will need an update, but that's ok to just update them and tag some versions.

Reply:
Thanks! That's what I'll do -- will make those proper bugs. This will mean that specifying a version on an interface will *require* a version specified on the package otherwise it will fail to parse.

- [ ] Make versioned `@since` in unversioned packages an error